Detectors/MUON/MID/Workflow/src/tracks-reader-workflow.cxx

```diff
@@ -0,0 +1,131 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file   MID/Workflow/src/tracks-reader-workflow.cxx
+/// \brief  DPL workflow to send MID tracks read from a root file
+/// \author Philippe Pillot, Subatech
+
+#include <algorithm>
+#include <cstring>
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <stdexcept>
+
+#include "Framework/runDataProcessing.h"
+#include "Framework/ControlService.h"
+#include "Framework/Task.h"
+
+#include "DPLUtils/RootTreeReader.h"
+
+#include "DataFormatsMID/ROFRecord.h"
+#include "DataFormatsMID/Track.h"
+
+using namespace o2::framework;
+using namespace o2::mid;
+
+class TrackSamplerTask
+{
+ public:
+  /// prepare the reader
+  void init(InitContext& ic)
+  {
+    auto inputFileName = ic.options().get<std::string>("infile");
+    mMinNumberOfROFsPerTF = ic.options().get<int>("repack-rofs");
+
+    mReader = std::make_shared<RootTreeReader>("midreco", inputFileName.c_str(), -1,
+                                               RootTreeReader::PublishingMode::Single,
+                                               RootTreeReader::BranchDefinition<std::vector<char>>{
+                                                 Output{"MID", "TRACKS", 0, Lifetime::Timeframe}, "MIDTrack"},
+                                               RootTreeReader::BranchDefinition<std::vector<ROFRecord>>{
+                                                 Output{"MID", "TRACKSROF", 0, Lifetime::Timeframe}, "MIDTrackROF"},
+                                               &mAccumulator);
+  }
+
+  /// process the next entry
+  void run(ProcessingContext& pc)
+  {
+    if (mReader->next()) {
```

Comment:
To avoid useless if...else, can't you just use:
while(mReader->next())

Reply:
The test is performed for every entry in the while loop anyway. Beside, I don't want to send everything in a single pair of messages, so I would need to loop only over the required number of entries to accumulate enough ROF and send the 2 messages, then let the DPL call the run() function again for the next TF. And add a test when there is nothing left to read to stop the DPL loop.